buffer/src/main/java/io/netty/buffer/api/internal/LeakDetection.java

```diff
@@ -0,0 +1,162 @@
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Utility class for the leak detection parts that are static and shared system-wide.
+ */
+@UnstableApi
+public final class LeakDetection {
+    static volatile int leakDetectionEnabled;
+
+    // Protected by synchronizing on the instance.
+    // This field is only accessed when leak are detected, or when callbacks are installed or removed.
+    private static final Map<Consumer<LeakInfo>, Integer> CALLBACKS = new IdentityHashMap<>();
+    private static final Integer INTEGER_ONE = 1;
+
+    private static final VarHandle LEAK_DETECTION_ENABLED_UPDATER;
+    static {
+        int enabled = SystemPropertyUtil.getBoolean("io.netty.buffer.leakDetectionEnabled", false) ? 1 : 0;
+        try {
+            LEAK_DETECTION_ENABLED_UPDATER = MethodHandles.lookup().findStaticVarHandle(
+                    LeakDetection.class, "leakDetectionEnabled", int.class);
+        } catch (Exception e) {
+            throw new ExceptionInInitializerError(e);
+        }
+        if (enabled > 0) {
+            CALLBACKS.put(LoggingLeakCallback.getInstance(), 1);
+        }
+        leakDetectionEnabled = enabled;
+    }
+
+    private LeakDetection() {
+    }
+
+    /**
+     * Internal API for {@link MemoryManager#onLeakDetected(Consumer)}.
+     *
+     * @see MemoryManager#onLeakDetected(Consumer)
+     */
+    public static SafeCloseable onLeakDetected(Consumer<LeakInfo> callback) {
+        requireNonNull(callback, "callback");
+        synchronized (CALLBACKS) {
+            Integer newValue = CALLBACKS.compute(callback, (k, v) -> v == null ? INTEGER_ONE : v + 1);
+            if (newValue.equals(INTEGER_ONE)) {
+                // This callback was not already in the map, so we need to increment the leak-detection-enabled counter.
+                LEAK_DETECTION_ENABLED_UPDATER.getAndAddAcquire(1);
+            }
+        }
+        return new CallbackRemover(callback);
+    }
+
+    /**
+     * Called when a leak is detected. This method will inform all registered
+     * {@linkplain MemoryManager#onLeakDetected(Consumer) on-leak-detected} callbacks.
+     *
+     * @param tracer The life-cycle trace of the leaked object.
+     * @param leakedObjectDescription A human-readable description of the leaked object, that can be used for logging.
+     */
+    public static void reportLeak(LifecycleTracer tracer, String leakedObjectDescription) {
+        requireNonNull(tracer, "tracer");
+        requireNonNull(leakedObjectDescription, "leakedObjectDescription");
+        synchronized (CALLBACKS) {
+            if (!CALLBACKS.isEmpty()) {
+                LeakInfo info = new InternalLeakInfo(tracer, leakedObjectDescription);
+                for (Consumer<LeakInfo> callback : CALLBACKS.keySet()) {
+                    callback.accept(info);
```

Comment:
Also it would be better if we could do the notification outside of the synchronised block.

Reply:
Then we have to copy. I think it's relatively rare that the set is modified, while the callbacks are only notified from the single Cleaner thread. The lock should be uncontended. To do both this and get cheaper iteration, we could do some copy-on-write thing with arrays. Do you think that's worth it?

Reply:
no... I was not thinking of the "only the cleaner thread" thingy. Fine as it is